src/term/mod.rs

```diff
@@ -200,63 +200,74 @@ pub use self::mode::TermMode;
 
 pub const TAB_SPACES: usize = 8;
 
-#[derive(Clone, Copy)]
-enum Charset {
-    Ascii,
-    /// A sparse charset only has a few characters in it and is best mapped with
-    /// a match statement
-    Sparse(StandardCharset),
-    /// A dense charset has most of the 96 possible characters in it and may be
-    /// better handler by a lookup table
-    Dense(StandardCharset),
-    /// A user defined charset handled by some more complicated mechanism
-    User,
+trait CharsetMapping {
+    fn map(&self, c: char) -> char;
 }
 
-impl Charset {
+impl CharsetMapping for StandardCharset {
     /// Switch/Map character to the active charset. Ascii is the common case and
     /// for that we want to do as little as possible.
     #[inline]
```

Comment:
This is a good idea, but in practice this won't be inlinable... unless we make the `fn input()` generic over `CharsetMapping`.